libvast/test/rest_api.cpp

```diff
@@ -0,0 +1,108 @@
+
+#include <vast/plugin.hpp>
+#include <vast/system/node.hpp>
+#include <vast/test/fixtures/actor_system_and_events.hpp>
+#include <vast/test/test.hpp>
+
+#include <regex>
+
+namespace {
+
+// "/query/{id}/next" -> "/query/:id/next"
+std::string to_express_format(const std::string& openapi_path) {
+  static const auto path_param = std::regex{"\\{(.+?)\\}"};
+  return std::regex_replace(openapi_path, path_param, ":$1");
+}
+
+struct fixture : public fixtures::deterministic_actor_system_and_events {
+  fixture()
+    : fixtures::deterministic_actor_system_and_events(
+      VAST_PP_STRINGIFY(SUITE)) {
+  }
+
+  vast::system::node_actor node;
+};
+
+class test_response final : public vast::http_response {
+public:
+  /// Append data to the response body.
+  void append(std::string body) override {
+    body_ += body;
+  }
+
+  /// Return an error and close the connection.
+  //  TODO: Statically verify that we can only abort
+  //  with the documented error codes.
+  void abort(uint16_t error_code, std::string message) override {
+    body_ = "";
+    error_
+      = caf::make_error(vast::ec::unspecified,
+                        fmt::format("http error {}: {}", error_code, message));
+  }
+
+  std::string body_ = {};
+  caf::error error_ = {};
+};
+
+} // namespace
+
+TEST(OpenAPI specs) {
+  for (auto const* rest_plugin :
+       vast::plugins::get<vast::rest_endpoint_plugin>()) {
+    MESSAGE("verifying spec for plugin " << rest_plugin->name());
+    auto endpoints = rest_plugin->api_endpoints();
+    auto spec = rest_plugin->openapi_specification();
+    REQUIRE(caf::holds_alternative<vast::record>(spec));
+    auto spec_record = caf::get<vast::record>(spec);
+    CHECK_EQUAL(endpoints.size(), spec_record.size());
+    for (auto const& [key, value] : spec_record) {
+      auto path = to_express_format(key);
+      auto endpoints_it = std::find_if(endpoints.begin(), endpoints.end(),
+                                       [&](auto const& endpoint) {
+                                         return endpoint.path == path;
+                                       });
+      REQUIRE(endpoints_it != endpoints.end());
+      // TODO: verify other endpoint parts
+    }
+  }
+}
+
+FIXTURE_SCOPE(rest_api_tests, fixture)
+
+TEST(status endpoint) {
+  auto const* plugin
+    = vast::plugins::find<vast::rest_endpoint_plugin>("api_status");
+  REQUIRE(plugin);
+  auto endpoints = plugin->api_endpoints();
+  REQUIRE_EQUAL(endpoints.size(), 1ull);
+  auto const& status_endpoint = endpoints[0];
+  auto handler = plugin->handler(self->system(), node);
+  auto response = std::make_shared<test_response>();
+  auto request = vast::http_request{
+    .params = {},
+    .response = response,
+  };
+  self->send(handler, vast::atom::http_request_v, status_endpoint.endpoint_id,
+             std::move(request));
+  run();
+  CHECK_EQUAL(response->error_, caf::error{});
+  CHECK(!response->body_.empty());
+  // TODO: Verify that the response contains valid YAML
```

Comment:
That can probably be done with `from_yaml`, which parses YAML strings into a `data`.

Reply:
Sadly, the comment is incorrect and it's JSON data instead of YAML.